### DISCUSSION TITLE
Provide more concise histogram shorthand

### DISCUSSION THREAD

**Discussion by chrispahm** [observablehq/plot#1982] [Ideas]
When working with one-dimensional data, I wish there was a very simple and easy to memorize way to generate a histogram. 
Something likes R's / matplotlib's `hist(data)`.

I know there's a Plot shorthand `Plot.rectY(data, Plot.binX()).plot()`, but it requires much more typing and is less ergenomic. Could something like `Plot.hist(data).plot()` be an alternative?

In an Observable notebook, what I usually end up doing is that I search and insert the "Histogram chart" snippet instead. The snippet requires an array of objects though, so the one-dimensional data needs to be transformed first. It would be nice to have a simpler solution for these scenarios!

**Comment by Fil** [comment]
This is what Plot.auto is for.

`Plot.auto(numbers, {x: numbers}).plot()`

(It would be nice if auto would know how to treat an array of numbers, allowing us to say ``Plot.auto(numbers).plot()`, but it's not yet the case.)

For a slightly different example with an accessor:
`Plot.auto({ length: 10000 }, { x: Math.random }).plot()`

**Reply by chrispahm** [reply]
If `Plot.auto(numbers).plot()` would work, that would be great!

Is there a more general reason why a simple check like `Array.isArray(firstArg) && firstArg.every(d => typeof x === "number")` 
(or something similar) wouldn't work to auto distinguish if the input is an array of numbers?

**Reply by chrispahm** [reply]
Guess this is the relevant section

https://github.com/observablehq/plot/blob/48b71d71988bfebe0228768f701c79a31b62ac67/src/marks/auto.js#L15-L17

so a simple check like this would probably be enough?
```js
export function autoSpec(data, options) {
  // Determine if data is an array of numbers.
  if (Array.isArray(data) && !options && data.every(d => typeof x === "number")) {
    options = { x: data }
  }
  options = normalizeOptions(options);

**Reply by Fil** [reply]
Yes I think we'd want something like that. We should also support typed arrays (those don't need a value check), as well as arrays of Date objects. Also usually we only detect the type of contents from the first non-nullish value, see for example
https://github.com/observablehq/plot/blob/48b71d71988bfebe0228768f701c79a31b62ac67/src/options.js#L412

**Reply by mbostock** [reply]
I feel like we could add

```js
Plot.histX(data, options?)
```

as shorthand for

```js
Plot.rectY(data, Plot.binX(undefined, options?))
```

I’m not sure it’s that useful, though… I imagine you’ll hit a wall pretty quickly which is primarily why we don’t offer this.

**Reply by chrispahm** [reply]
Thanks for the clarification! I think it would be great if `Plot.auto` would support an array of numbers like @Fil described. I updated the snippet above to include his suggestions, so probably this little change would be sufficient?

```js
// ...
import {isColor, isNumeric, isObject, isOptions, isOrdinal, isTemporal, labelof, valueof} from "../options.js";
// ...

export function autoSpec(data, options) {
  // Determine if data is an array of numbers or dates.
  if (Array.isArray(data) && !options && (ArrayBuffer.isView(data) || isNumeric(data) || isTemporal(data)) ) {
    options = { x: data };
  }
  options = normalizeOptions(options);
  ```

If desired I can create a PR as well.
